Introduce make targets and charts for the development, and deployment of `grove-kube-scheduler`. (#80)
NVIDIA/grove: docs/developing-scheduler.md

> # developing and testing the grove kube-scheduler plugin

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).

Testing the grove kube-scheduler plugin is currently supported locally using a kind cluster. There are make targets present in the [scheduler/Makefile](../scheduler/Makefile) that support this.

Likewise, the scheduler can be tested in any kubernetes cluster by deploying the kube-scheduler built with the grove plugin as a pod.

> ## building the scheduler

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).

To build the kube-scheduler image, use the `docker-build-<arch>` target. The image built from here can be deployed on any kubernetes cluster for testing.
This step is *NOT NECESSARY* for testing the scheduler locally in the kind cluster, as it is taken care of by skaffold.

> ## testing the scheduler locally

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).

You can test the scheduler locally in two ways:

> - Replace the vanilla kube-scheduler that a kind cluster starts with with the grove-kube-scheduler.

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).

> - Deploy the grove-kube-scheduler as a second scheduler running alongside the vanilla kube-scheduler.

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).

> ### replacing the vanilla kube-scheduler

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).

> - Bring up a kind cluster through the specified target `kind-up` in [scheduler/Makefile](../scheduler/Makefile). (wait for the node to be ready)

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).

> - Modify the `KubeSchedulerConfigruation` according to your needs in `generate_kube_scheduler_configuration` function in [replace-scheduler.sh](../scheduler/hack/replace-scheduler.sh) if necessary. This function generates the configuration that you want to pass to the kube-scheduler with the grove plugin; and is later copied to the kind node.

Author: Addressed in [bcd27ea](https://github.com/NVIDIA/grove/pull/80/commits/bcd27eaa06bab5e5d2fddfbb97142046d6f71f44).

> - Modify the static pod manifest in `generate_kube_scheduler_manifest` function in [replace-scheduler.sh](../scheduler/hack/replace-scheduler.sh) if necessary. This function generates the static pod manfiest of the kube-scheduler pod that will be replacing the vanilla kube-scheduler in the kind cluster.

Author: Addressed in [bcd27ea](https://github.com/NVIDIA/grove/pull/80/commits/bcd27eaa06bab5e5d2fddfbb97142046d6f71f44).

> - Run `make deploy-local-replace`.

Reviewer: `make deploy-local-replace` has too many verbs in it - `deploy`, `replace` which is not very clear.
Possible options that come to mind at the time of reviewing the PR:
* make replace-default-scheduler
* make swap-default-scheduler

Author: Addressed in [bcd27ea](https://github.com/NVIDIA/grove/pull/80/commits/bcd27eaa06bab5e5d2fddfbb97142046d6f71f44).

Following the above should replace the vanilla kube-scheduler with grove-kube-scheduler.
To iterate on changes, make changes in the [scheduler](../scheduler), and rerun the make target.

> ### running as a second kube-scheduler

Reviewer: ```suggestion
### Running as a second kube-scheduler
```

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).

> - Bring up a kind cluster through the specified target `kind-up` in [scheduler/Makefile](../scheduler/Makefile). (wait for the node to be ready)

Reviewer: ```suggestion
- Bring up a kind cluster through the specified target `kind-up` in [scheduler/Makefile](../scheduler/Makefile). Wait for the node to be ready.
```

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).

> - Modify the [values.yaml](../scheduler/charts/values.yaml) and the corresponding [templates](../scheduler/charts/templates).

Reviewer: Modify what? Is it necessary or the default set of values and templates are sufficient?

Author: The default set of values is sufficient for running the scheduler. Further modifications to the configuration if needed can be made in values.yaml. Will clarify this.

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).

> - Run `make deploy-local-second`.

Reviewer: Possible alternatives:
`make deploy-second-scheduler`
`make deploy-custom-scheduler`

Author: Addressed in [bcd27ea](https://github.com/NVIDIA/grove/pull/80/commits/bcd27eaa06bab5e5d2fddfbb97142046d6f71f44).

> Following the above should deploy grove-kube-scheduler as a second scheduler in the cluster. Specify the `schedulerName` to make use of the second scheduler.

Reviewer: ```suggestion
Following the above should deploy grove-kube-scheduler as a second scheduler in the cluster. Specify the `schedulerName` in the `PodSpec.SchedulerName` to ensure that the grove scheduler is used to schedule the target pod(s).
```

Author: Addressed in [6cd5262](https://github.com/NVIDIA/grove/pull/80/commits/6cd5262cc49b7c3deeb7515b2393d8590569e559).
